### DISCUSSION TITLE
Spec for player-buildable 'randomised' vehicles.

### DISCUSSION THREAD

**Discussion by andythenorth** [OpenTTD/OpenTTD#12299] [Features]
Iron Horse includes 'randomised wagons'.

<img width="506" alt="image" src="https://github.com/OpenTTD/OpenTTD/assets/1780327/0685e6d9-9539-4716-8018-01c83c9f0420">

These are vehicles that use an action 0 ID in grf, and then for graphics, chain to a randomly selected varact2 chain of another vehicle (from a pre-defined candidate list), which terminates in that vehicle's action 1 realsprites.

The Iron Horse implementation also tunes the 'random' weights, so that sequences of similar wagons appear in short runs, as this looks better. 

The Iron Horse implementation is mostly orientated to building 'random' within a variant group, but also has a set of 'random' vehicles that transcend multiple variant groups.

These work fine, have good results, and work with autoreplace

But they also
- are mildly convoluted in the implementation
- consume concurrent action 2 IDs rapidly, against the limit of 255, as they sometimes span across many action 0-3-2-1 blocks for vehicles
- are specific to Iron Horse and depend on Iron Horse compile features, whereas this might be a feature that is useful to make more widely available to grf authors

So this is a discussion to ask: what might be useful if the grf spec provided support for 'randomised' vehicles?

A callback providing a list of vehicle IDs, and then OpenTTD builds an actual vehicle from the list?  

Leave it up to individual grf authors (and perhaps increase the concurrent action 2 ID limit)?

Something else?

**Comment by 2TallTyler** [comment]
I have a proposal for this already, albeit not a patch yet. 😛 

1. Add a Ctrl+Click action on the `Buy Vehicle` button to purchase a random variant of the currently-selected vehicle, at the same depth, if available. If no variants are available, that vehicle is purchased — effectively ignoring the Ctrl modifier.
2. Add a boolean vehicle property `random_variant` that indicates if the vehicle was purchased randomly. This flag is checked when autoreplacing or cloning, and if present the new vehicle is randomized in the same manner.

Pros:
* No GRF spec addition needed, uses the existing variant mechanic
* No new buttons
* Flexible for GRF authors to do as they please

Cons:
* Random flag is invisible to players and cannot be un-set manually. This may cause unexpected behavior, particularly if a player autoreplaces a vehicle from one GRF to another, that uses variants in a different way (liveries versus length, for example). This could be mitigated with a GRF spec flag that enables randomization.

**Reply by reldred** [reply]
I like this, would certainly handle the random vehicles in Iron Horse except the variants that provide vehicles with different specs (Half of a normally coupled dual head arrangement, for instance), and wouldn't handle some of Andy's wagons where he has a randomized variation that selects from multiple separate variant sets like box cars and flat wagons.
